File: core/src/main/java/org/maxur/mserv/web/grizzly/RuntimeExceptionHandler.java
/*
 * Copyright 2015 Russian Post
 *
 * This source code is Russian Post Confidential Proprietary.
 * This software is protected by copyright. All rights and titles are reserved.
 * You shall not use, copy, distribute, modify, decompile, disassemble or reverse engineer the software.
 * Otherwise this violation would be treated by law and would be subject to legal prosecution.
 * Legal use of the software provides receipt of a license from the right holder only.
 */


package org.maxur.mserv.web.grizzly;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.ws.rs.core.GenericEntity;
import javax.ws.rs.core.Response;
import javax.ws.rs.ext.ExceptionMapper;
import java.util.List;

import static javax.ws.rs.core.MediaType.APPLICATION_JSON;
import static javax.ws.rs.core.Response.Status.INTERNAL_SERVER_ERROR;

/**
 * The type Runtime exception handler.
 *
 * @author myunusov
 * @version 1.0
 * @since <pre>11.09.2015</pre>
 */
public class RuntimeExceptionHandler implements ExceptionMapper<RuntimeException> {
    private static final Logger LOGGER = LoggerFactory.getLogger(RuntimeExceptionHandler.class);

    @Override
    public Response toResponse(RuntimeException exception) {
        LOGGER.error(exception.getMessage(), exception);
        return Response
                .status(INTERNAL_SERVER_ERROR)
                .type(APPLICATION_JSON)
                .entity(makeErrorEntity(exception))
                .build();
    }

    private GenericEntity<List<Incident>> makeErrorEntity(final RuntimeException exception) {
        return new GenericEntity<List<Incident>>(Incident.incidents("System error", exception.getMessage())) {
        };
    }


}

File: maxur-mserv-core/src/main/kotlin/org/maxur/mserv/core/rest/RuntimeExceptionHandler.kt
package org.maxur.mserv.core.rest

import org.slf4j.Logger
import org.slf4j.LoggerFactory
import javax.ws.rs.WebApplicationException
import javax.ws.rs.core.GenericEntity
import javax.ws.rs.core.MediaType.APPLICATION_JSON
import javax.ws.rs.core.Response
import javax.ws.rs.core.Response.Status
import javax.ws.rs.core.Response.status
import javax.ws.rs.ext.ExceptionMapper

/**
 * The type Runtime exception handler.
 * @author myunusov
 * *
 * @version 1.0
 * *
 * @since <pre>11.09.2015</pre>
 */
class RuntimeExceptionHandler : ExceptionMapper<RuntimeException> {

    companion object {
        val log: Logger = LoggerFactory.getLogger(RuntimeExceptionHandler::class.java)
    }

    override fun toResponse(exception: RuntimeException): Response {
        when (exception) {
            is IllegalArgumentException -> return onIllegalArgument(exception)
            is WebApplicationException -> return onWebApplicationException(exception)
            else -> {
                log.error(exception.message, exception)
                return status(Status.INTERNAL_SERVER_ERROR)
                        .type(APPLICATION_JSON)
                        .entity(makeSystemErrorEntity(exception))
                        .build()
            }
        }
    }

    private fun onWebApplicationException(exception: WebApplicationException): Response {
        log.debug(exception.message, exception)
        return exception.response
    }

    private fun onIllegalArgument(exception: IllegalArgumentException): Response {
        log.warn(exception.message)
        log.debug(exception.message, exception)
        return status(Status.BAD_REQUEST)
                .type(APPLICATION_JSON)
                .entity(makeUserErrorEntity(exception))
                .build()
    }

    private fun makeSystemErrorEntity(exception: RuntimeException): GenericEntity<List<Incident>> {
        return object : GenericEntity<List<Incident>>(
                Incident.incidents("System error", exception.message ?: "Unknown")
        ) {
        }
    }

    private fun makeUserErrorEntity(exception: RuntimeException): GenericEntity<List<Incident>> {
        return object : GenericEntity<List<Incident>>(
                Incident.incidents("Bad parameters", exception.message ?: "Unknown")
        ) {

        }
    }

}
